src/order/filter/filter_product.lean

```diff
@@ -0,0 +1,223 @@
+/-
+Copyright (c) 2019 Abhimanyu Pallavi Sudhir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Abhimanyu Pallavi Sudhir
+"Filterproducts" (ultraproducts on general filters), ultraproducts and 
+the construction of the hyperreals.
+-/
+
+import data.real.basic
+import order.filter.basic
+
+universes u v
+variables {α : Type u} (β : Type v) (φ : filter α)
+local attribute [instance] classical.prop_decidable
+
+namespace filter 
+
+/-- Two sequences are bigly equal iff the kernel of their difference is in φ -/
+def bigly_equal : setoid (α → β) := 
+⟨ λ a b, {n | a n = b n} ∈ φ,
+  λ a, by simp only [eq_self_iff_true, (set.univ_def).symm, univ_sets], 
+  λ a b ab, by simpa only [eq_comm], 
+  λ a b c ab bc, sets_of_superset φ (inter_sets φ ab bc) 
+    (λ n (r : a n = b n ∧ b n = c n), eq.trans r.1 r.2)⟩
+
+/-- Ultraproduct, but on a general filter -/
+def filterprod := quotient (bigly_equal β φ)
+local notation `β*` := filterprod β φ
+
+namespace filter_product
+
+def seq_to_filterprod : (α → β) → β* := @quotient.mk' (α → β) (bigly_equal β φ)
+
+/-- Equivalence class containing the constant sequence of a term in β -/
+def to_filterprod : β → β* := function.comp (seq_to_filterprod β φ) (λ x n, x)
+
+theorem to_filterprod_inj (NT : φ ≠ ⊥): function.injective (to_filterprod β φ) :=
+begin
+  intros r s rs, by_contra N,
+  rw [to_filterprod, seq_to_filterprod, quotient.eq', bigly_equal] at rs, 
+  simp only [N, set.set_of_false, empty_in_sets_eq_bot] at rs,
+  exact NT rs
+end
+
+instance : has_coe (α → β) β* := ⟨ seq_to_filterprod β φ ⟩
+
+instance coe_filterprod : has_coe β β* := ⟨ to_filterprod β φ ⟩
+
+instance [has_add β] : has_add β* :=
+{ add := λ x y, quotient.lift_on₂' x y (λ a b, (quotient.mk' $ λ n, a n + b n : β*)) $
```

Comment:
This is an instance of a general theorem `lift : (β -> β) -> (β* -> β*)` and `lift2 : (β -> β -> β) -> (β* -> β* -> β*)`. @johoelzl do you see a way to generalize both of these into a single lift function?

Reply:
I've made general definitions `lift`, `lift₂`, `lift_rel`, `lift_rel₂` -- I guess if there was a way to generalise this, it would require generalising also `quotient.lift_on₂`, etc.